File: practice/src/log_polar.cpp
// This program captures frames from camera and stores
// them in a file in log polar form
//
// author: Anshay
// Date: 06/01/2021

#include <opencv4/opencv2/opencv.hpp>
#include <iostream>
#include <string>

int main(int argc, char** argv){
    cv::VideoCapture cap;
    if(argc==1){
        std::cout<<"\nNot enough arguments provided!\n"
                    "Usage: log_polar <output_file_path> "
                    "(optional)<input_file_path>\n";
        return 1;
    }
    if(argc==2){
        cap.open(0);
    }
    else{
        cap.open(std::string(argv[2]));
    }
    if(!cap.isOpened()){
        std::cout<<"\nCould not open file/camera\n";
        exit(0);
    }
    double fps = cap.get(cv::CAP_PROP_FPS);
    cv::Size size((int)cap.get(cv::CAP_PROP_FRAME_WIDTH), (int)cap.get(cv::CAP_PROP_FRAME_HEIGHT));

    cv::VideoWriter writer;
    writer.open(argv[1], cv::VideoWriter::fourcc('M','J','P','G'), fps, size);

    std::string WINDOW_NAME = "Log Polar";
    cv::namedWindow(WINDOW_NAME);
    cv::Mat frame;
    while(true){
        cap >> frame;
        cv::logPolar(frame, frame, cv::Point2f(frame.cols/2, frame.rows/2), 40, cv::WARP_FILL_OUTLIERS);
        cv::imshow(WINDOW_NAME, frame);
        writer << frame;
        char c = cv::waitKey(33);
        if(c==27) break;
    }
    writer.release();
    std::cout<<"\nCapture complete\n";
    return 0;
}

File: practice/scripts/log_polar.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


# This program captures frames from camera and stores
# them in a file in log polar form
# 
# author: Anshay
# Date: 06/01/2021


# In[2]:


import numpy as np
import cv2
import sys


# In[13]:


def log_polar(FILENAME):
    window_name = "Log Polar"
    cv2.namedWindow(window_name, cv2.WINDOW_NORMAL)
    cap = cv2.VideoCapture()
    writer = cv2.VideoWriter()
    cap.open(0)
    if(not cap.isOpened()):
        print("\nCould not open camera")
        exit()
    try:
        ret, frame = cap.read()
        fps = 24.0
        size = (frame.shape[1],frame.shape[0])
        writer.open(FILENAME, cv2.VideoWriter_fourcc(*"MJPG"), fps, size)
        while(ret):
            frame = cv2.warpPolar(frame, size, (size[0]/2, size[1]/2), 40, cv2.WARP_POLAR_LINEAR)
            cv2.imshow(window_name, frame)
            writer.write(frame)
            if(cv2.waitKey(33)==27):
                print("\nExiting and closing camera")
                break
            ret, frame = cap.read()
    finally:
        writer.release()
        cap.release()


# In[14]:


if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("\nCorrect usage: capture_and_store <output_file_path>\n")
        exit()
    filename = sys.argv[1]
    log_polar(filename)
    cv2.destroyAllWindows()


# In[ ]:




